artiq/gateware/rtio/phy/edge_counter.py

```diff
@@ -0,0 +1,79 @@
+from migen import *
+from artiq.gateware.rtio import rtlink
+
+
+class SimpleEdgeCounter(Module):
+    """Counts rising/falling edges of an input signal.
+
+    Control (sensitivity/zeroing) is done via a single RTIO output channel,
+    which is is also used to request an input event to be emitted with the
+    current counter value.
+
+    :param input_state: The (scalar) input signal to detect edges of. This
+        should already be in the rio_phy clock domain.
+    :param counter_width: The width of the counter register, in bits. Defaults
+        to 31 to match integers being signed in ARTIQ Python.
+    """
+
+    def __init__(self, input_state, counter_width=31):
```

Comment:
Not saying that this is wrong, just trying to understand the rationale: Is the reason that you use 31 here just that it makes both interpreting it as a u32 and upcasting to i64 difficult?
Your testbench uses 32 bits.